sourcetable.go

```diff
@@ -128,6 +131,44 @@ func (m StreamEntry) String() string {
 		m.Authentication, fee, m.Bitrate, m.Misc)
 }
 
+// GetSourcetable fetches a source table from a specific caster.
+//
+// The funciton returns a list of errors which can be treated as warnings.
+// These warnings indicate that the caster is returning an improper rtcm3 format.
+func GetSourcetable(ctx context.Context, url string) (Sourcetable, []error, error) {
+	warnings := []error{}
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+	if err != nil {
+		return Sourcetable{}, warnings, errors.Wrap(err, "building request")
+	}
+
+	req.Header.Set("Ntrip-Version", "Ntrip/2.0")
+	req.Header.Set("User-Agent", "ntrip-mqtt-gateway")
+
+	client := &http.Client{}
+
+	res, err := client.Do(req)
+	if err != nil {
+		return Sourcetable{}, warnings, err
+	}
+	defer res.Body.Close()
+
+	body, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		return Sourcetable{}, warnings, err
+	}
+
+	if res.StatusCode != 200 {
+		return Sourcetable{}, warnings, fmt.Errorf("received a non 200 status code")
+	}
+
+	// Swollowing the errors here is okay because the errors are more like warnings.
+	// All rows that could be parsed will be present in the source table.
+	table, warnings := ParseSourcetable(string(body[:]))
```

Comment:
It's a little weird that if the response body is not at all a sourcetable (just some other content entirely) the error will be nil (with a bunch of parse errors).

It's definitely more convenient for us to do it this way, so we're not stopped from parsing the information we care about if the first line is a malformed CAS entry or whatever.

Reply:
I agree kind of. I considered three signatures for the ParseSourcetable (not returning any errors, returning a list of warnings, and returning a list of warnings and an error). I think the last option would handle the situation where you got back random data better, but I thought it added complexity for a use case that would almost never occur.

Happy to change it if you think that is better.

Reply:
No, that's all good. I agree it's probably not necessary. Happy to go with this approach.

The only other approach that I think would be reasonable would be much more heavy-handedly returning an error as soon as you get a parsing error. Where the rationale is that you Get a valid Sourcetable, or you error - and if the Sourcetable contains any invalid lines then it is invalid.

Again, this is good.

Reply:
Awesome. I'm going to move back to gnss-stream for a bit.